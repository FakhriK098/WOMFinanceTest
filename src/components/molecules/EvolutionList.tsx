import React, { useCallback } from 'react';
import { FlatList, Image, Pressable, StyleSheet } from 'react-native';
import Text from '../atoms/Text';
import type { EvolutionEntry } from '@typings/pokemon';
import { spacing } from '@theme/spacing';
import { radius } from '@theme/radius';
import { colors } from '@theme/colors';
import { typography } from '@theme/typography';

type Props = {
  items: EvolutionEntry[];
  onPressItem: (id: number) => void;
};

const EvolutionList = ({ items, onPressItem }: Props) => {
  const renderItem = useCallback(
    ({ item }: { item: EvolutionEntry }) => (
      <Pressable style={styles.card} onPress={() => onPressItem(item.id)}>
        {!!item.image && (
          <Image
            source={{ uri: item.image }}
            style={styles.img}
            resizeMode="contain"
          />
        )}
        <Text size={typography.xs} color={colors.text.light}>
          {item.name}
        </Text>
      </Pressable>
    ),
    [onPressItem],
  );
  return (
    <FlatList
      data={items}
      horizontal
      keyExtractor={it => String(it.id)}
      renderItem={renderItem}
      contentContainerStyle={styles.container}
      showsHorizontalScrollIndicator={false}
    />
  );
};

const styles = StyleSheet.create({
  container: {
    gap: spacing.sm,
  },
  card: {
    width: 100,
    alignItems: 'center',
    padding: spacing.sm,
    backgroundColor: '#FFFFFF',
    borderRadius: radius.md,
  },
  img: {
    width: 72,
    height: 72,
    marginBottom: spacing.xs,
  },
});

export default React.memo(EvolutionList);
