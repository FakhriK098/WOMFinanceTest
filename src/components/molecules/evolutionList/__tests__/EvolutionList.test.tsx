import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { FlatList } from 'react-native';

import EvolutionList from '../EvolutionList';

const items = [
  { id: 1, name: 'Bulbasaur', image: 'http://img1' },
  { id: 2, name: 'Ivysaur', image: 'http://img2' },
  { id: 3, name: 'Venusaur', image: 'http://img3' },
];

describe('EvolutionList molecule', () => {
  it('matches snapshot', () => {
    let tr: renderer.ReactTestRenderer;
    act(() => {
      tr = renderer.create(
        <EvolutionList items={items as any} onPressItem={() => {}} />,
      );
    });
    expect(tr!.toJSON()).toMatchSnapshot();
    act(() => tr!.unmount());
  });

  it('calls onPressItem with id', () => {
    const onPress = jest.fn();
    let tr: renderer.ReactTestRenderer;
    act(() => {
      tr = renderer.create(
        <EvolutionList items={items as any} onPressItem={onPress} />,
      );
    });
    // Render a single item via FlatList's renderItem and trigger onPress
    const list = tr!.root.findByType(FlatList);
    const element = list.props.renderItem({ item: items[0], index: 0 });
    let itemRenderer: renderer.ReactTestRenderer;
    act(() => {
      itemRenderer = renderer.create(element);
    });
    const pressable = itemRenderer.root.find((inst) => inst.props && typeof inst.props.onPress === 'function');
    act(() => {
      pressable.props.onPress();
    });
    expect(onPress).toHaveBeenCalledWith(1);
    act(() => itemRenderer.unmount());
    act(() => tr!.unmount());
  });
});
