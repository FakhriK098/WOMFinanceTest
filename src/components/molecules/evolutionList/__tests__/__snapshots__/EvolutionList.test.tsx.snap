// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EvolutionList molecule matches snapshot 1`] = `
<RCTScrollView
  contentContainerStyle={
    {
      "gap": 8,
    }
  }
  data={
    [
      {
        "id": 1,
        "image": "http://img1",
        "name": "Bulbasaur",
      },
      {
        "id": 2,
        "image": "http://img2",
        "name": "Ivysaur",
      },
      {
        "id": 3,
        "image": "http://img3",
        "name": "Venusaur",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={true}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  showsHorizontalScrollIndicator={false}
  stickyHeaderIndices={[]}
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={
        [
          {
            "flexDirection": "row",
          },
          null,
        ]
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderRadius": 12,
            "padding": 8,
            "width": 100,
          }
        }
      >
        <Image
          resizeMode="contain"
          source={
            {
              "uri": "http://img1",
            }
          }
          style={
            {
              "height": 72,
              "marginBottom": 4,
              "width": 72,
            }
          }
        />
        <Text
          style={
            [
              {
                "color": "#111827",
                "fontSize": 12,
                "fontWeight": "400",
              },
              undefined,
            ]
          }
        >
          Bulbasaur
        </Text>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={
        [
          {
            "flexDirection": "row",
          },
          null,
        ]
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderRadius": 12,
            "padding": 8,
            "width": 100,
          }
        }
      >
        <Image
          resizeMode="contain"
          source={
            {
              "uri": "http://img2",
            }
          }
          style={
            {
              "height": 72,
              "marginBottom": 4,
              "width": 72,
            }
          }
        />
        <Text
          style={
            [
              {
                "color": "#111827",
                "fontSize": 12,
                "fontWeight": "400",
              },
              undefined,
            ]
          }
        >
          Ivysaur
        </Text>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={
        [
          {
            "flexDirection": "row",
          },
          null,
        ]
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderRadius": 12,
            "padding": 8,
            "width": 100,
          }
        }
      >
        <Image
          resizeMode="contain"
          source={
            {
              "uri": "http://img3",
            }
          }
          style={
            {
              "height": 72,
              "marginBottom": 4,
              "width": 72,
            }
          }
        />
        <Text
          style={
            [
              {
                "color": "#111827",
                "fontSize": 12,
                "fontWeight": "400",
              },
              undefined,
            ]
          }
        >
          Venusaur
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
